feat: run unit tests per fork

amkCha opened the pull request:
A flow suggestion on how to run unit tests per fork 
- set an env variable `UNIT_REPLAY_TESTS_FORK`
- add a CI Action for Shanghai Unit tests that runs in // of current Unit London and Replay
- set a property `unit.replay.tests.fork` that copies the env variable or default to `LONDON` in gradle
- leverage `TracerTestBase` class setup by Gaurav and from which all tests inherit
- use gradle property set above to add a `chainConfig` to the bare `testInfo` previously available in `TracerTestBase`
- pass `testInfoWithChainConfig` (didn't rename it for now) to the `ToyExecutionEnvironment.build`
- build tracer from the fork available in the chain config

amkCha committed:
feat: testInfoWithChainConfig in builder toyExecEnv

amkCha committed:
feat: remove dependence on chain id

amkCha committed:
fix: header

amkCha committed:
feat: fix return test and rebase

amkCha committed:
feat: set replay to London fork

amkCha committed:
feat: multiblock testinfo

amkCha commented on testing/src/main/java/net/consensys/linea/reporting/TracerTestBase.java:
```diff
@@ -14,14 +14,28 @@
 
 public class TracerTestBase {
-  public TestInfo testInfo;
+  public static TestInfoWithChainConfig testInfo = new TestInfoWithChainConfig();
```
Can be renamed :ok_hand: , it was easier to keep `testInfo` to change all the tests while we agree on the implementation

amkCha commented on testing/src/main/java/net/consensys/linea/testing/BytecodeCompiler.java:
```diff
@@ -45,7 +45,7 @@ private BytecodeCompiler() {}
    */
   public static BytecodeCompiler newProgram() {
-    loadOpcodes(UNIT_TEST_CHAIN.fork);
+    loadOpcodes(Fork.LONDON);
```
To do: make dynamic

amkCha committed:
feat: refacto Bytecode dynamic loadOpcodes

amkCha commented on arithmetization/src/test/java/net/consensys/linea/zktracer/exceptions/multiExceptions/ReturnTest.java:
```diff
@@ -108,10 +109,13 @@ void maxCodeSizeAndOogExceptionForCreate() {
+    // + ((32027-3-3-6-3-3-3-32000/64))(less than 0.5 so not adding gas) + 3L PUSH + 3L PUSH
     // 21000L for the intrinsic transaction cost
-    bytecodeRunner.run(32027L + 21000L, testInfo);
+    var initCodeCost = isPostShanghai(testInfo.chainConfig.fork) ? 2L : 0L;
```
@DavePearce an example of what we talked about earlier where gas is hardcoded and different when running on London and Shanghai - this is a hotfix while we find a better way to account for the difference :+1:

DavePearce commented on .github/workflows/gradle-tests.yml:
```diff
@@ -51,9 +51,9 @@ jobs:
+  # Unit Tests for London
   # ==================================================================
-  unit-tests:
+  unit-tests-london:
```
I wonder if its possible write a reusable action here to avoid duplicating everything, especially as some of the settings for memory usage etc, are tricky?

amkCha replied:
Good idea, let me try to do that !

DavePearce reviewed (approved):
Overall, looks good to me.  The only thing I really didn't like was the name of the environment variable `UNIT_REPLAY_TESTS_FORK`.  I don't think that's a good choice of name, as it implies its only for use with the unit/replay tests.  I would just use `ZKEVM_FORK` and then we can reuse the same variable across e.g. replay tests, blockchain tests, whatever.

The other comment I have, which is just to give you an idea of what I might have done differently.  You have a lot of things like this:

```
final ToyExecutionEnvironmentV2 executionEnvironmentV2 =
        ToyExecutionEnvironmentV2.builder(testInfo)
```

and e.g.

```
final Bytes bytecode =
        BytecodeCompiler.newProgram(testInfo)
```

Now, given that all the tests extend `TraceTestBase`, I would probably have taken the opportunity to hide some of this configuration.  It doesn't save time for this PR though, but it might in future PRs (then again, it might not).  Roughly, I would have added methods like this to `TracerTestBase`:

```
protected BytecodeCompiler newBytecodeProgram() {
   return BytecodeCompiler.newProgram(testInfo);
}
```

And, something similar for the `ToyExecutionEnvironmentV2`, etc.  Then, we would call these methods in our tests instead of calling e.g. `BytecodeCompiler.newProgram(testInfo)` directly, etc.  It means if we want to make future changes regarding configurations like this, we might only have to change these methods rather than all the places they are called from.  

Anyways, I wouldn't actually make the change at this stage, since you've got the PR ready to go.

DavePearce committed:
fix outstanding compilation errors

This fixes the outstanding compilation errors, such that the trace now
compiles and passes at least one test.  However, I believe there are
still problems related to the use of testInfo in static variable
initialisers.  Specifically, we will see that it is null in such
situations, leading to some kind of exception.

amkCha committed:
feat: parameterize unit tests

amkCha committed:
fix: RevertingLogsTests

amkCha commented:
> Overall, looks good to me. The only thing I really didn't like was the name of the environment variable `UNIT_REPLAY_TESTS_FORK`. I don't think that's a good choice of name, as it implies its only for use with the unit/replay tests. I would just use `ZKEVM_FORK` and then we can reuse the same variable across e.g. replay tests, blockchain tests, whatever.
> 
> The other comment I have, which is just to give you an idea of what I might have done differently. You have a lot of things like this:
> 
> ```
> final ToyExecutionEnvironmentV2 executionEnvironmentV2 =
>         ToyExecutionEnvironmentV2.builder(testInfo)
> ```
> 
> and e.g.
> 
> ```
> final Bytes bytecode =
>         BytecodeCompiler.newProgram(testInfo)
> ```
> 
> Now, given that all the tests extend `TraceTestBase`, I would probably have taken the opportunity to hide some of this configuration. It doesn't save time for this PR though, but it might in future PRs (then again, it might not). Roughly, I would have added methods like this to `TracerTestBase`:
> 
> ```
> protected BytecodeCompiler newBytecodeProgram() {
>    return BytecodeCompiler.newProgram(testInfo);
> }
> ```
> 
> And, something similar for the `ToyExecutionEnvironmentV2`, etc. Then, we would call these methods in our tests instead of calling e.g. `BytecodeCompiler.newProgram(testInfo)` directly, etc. It means if we want to make future changes regarding configurations like this, we might only have to change these methods rather than all the places they are called from.
> 
> Anyways, I wouldn't actually make the change at this stage, since you've got the PR ready to go.

I renamed the env variable to `ZKEVM_FORK` :ok_hand: 
And I think you are right, we should leverage `TracerTestBase` a bit more but might need more time to understand what should go where. So I'll merge this and we can refactor it afterwards. For instance, the issue you brought up with the test `RevertingLogsTests` not lauching locally because `testInfo` is initialized only when tests starts (and it was called outside of a test so it was failing) makes me think that we could actually have a task to store the `chainConfig` in a class and run this task prior to all gradle test tasks - tbd

amkCha committed:
fix: ZkCounter extends TracerTestBase
